pkg/dispatcher/dispatcher.go

```diff
@@ -0,0 +1,204 @@
+		},
+	}
+	instance, err := d.instanceDao.UpSert(ctx, instance)
+	if err != nil {
+		return fmt.Errorf("unable to create maestro instance: %s", err.Error())
+	}
+	logger.Infof("Created maestro instance %s in database", instance.Name)
+
+	instances, err := d.instanceDao.All(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to list maestro instances: %s", err.Error())
+	}
+
+	// initialize the consistent hash ring
+	members := []consistent.Member{}
+	for _, instance := range instances {
+		if instance.UpdatedAt.After(time.Now().Add(time.Duration(-1*d.instanceExpirationPeriod) * time.Second)) {
+			members = append(members, instance)
+		}
+	}
+
+	d.consistent = consistent.New(members, consistent.Config{
+		PartitionCount:    7,
+		ReplicationFactor: 20,
+		Load:              1.25,
+		Hasher:            hasher{},
+	})
+	logger.Infof("Initialized consistent hash ring with members: %v", members)
+
+	consumers, err := d.consumerDao.All(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to list consumers: %s", err.Error())
+	}
+
+	// initialize the consumer set for current instance
+	for _, consumer := range consumers {
+		instanceID := d.consistent.LocateKey([]byte(consumer.ID)).String()
+		if instanceID == d.instanceID {
+			// new consumer added to the current instance, need to resync resource status updates for this consumer
+			d.consumerSet.Add(consumer.ID)
+			logger.Infof("Added new consumer %s to consumer set for instance %s", consumer.ID, d.instanceID)
+			d.resyncChan <- consumer.ID
+		}
+	}
+	logger.Infof("Initialized consumers %s for current instance %s", d.consumerSet.String(), d.instanceID)
+
+	pulseTicker := time.NewTicker(time.Duration(d.pulseInterval) * time.Second)
+	checkTicker := time.NewTicker(time.Duration(d.checkInterval) * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			pulseTicker.Stop()
+			checkTicker.Stop()
+			close(d.resyncChan)
+			return nil
+		case <-pulseTicker.C:
+			logger.Infof("Updating heartbeat for maestro instance: %s", instance.Name)
+			// update the heartbeat for current instance
+			instance.UpdatedAt = time.Now()
+			instance, err = d.instanceDao.Replace(ctx, instance)
+			if err != nil {
+				return fmt.Errorf("unable to update heartbeat for maestro instance: %s", err.Error())
+			}
+		case <-checkTicker.C:
+			logger.Infof("Checking maestro instances and updating the hash ring")
+			// check for instances that are not updated in the last instanceExpirationPeriod seconds
+			instances, err = d.instanceDao.All(ctx)
+			if err != nil {
+				return fmt.Errorf("unable to list maestro instances: %s", err.Error())
+			}
+			for _, instance := range instances {
+				if instance.UpdatedAt.After(time.Now().Add(time.Duration(-1*d.instanceExpirationPeriod) * time.Second)) {
+					d.consistent.Add(instance)
+				} else {
+					d.consistent.Remove(instance.Name)
+				}
+			}
+			logger.Infof("members in consistent hash ring: %v", d.consistent.GetMembers())
+			consumers, err = d.consumerDao.All(ctx)
```

Comment:
can you add a TODO here? this could be heavy and we have to do it periodically. Do we know how it performs?

Reply:
Added a TODO item.
As an alternative, we might consider rebalancing only those consumers whose owner (Maestro instance) is inactive or deceased.
Another scenario involves the creation of new consumers, where we can leverage the controller manager handler to update the consumer set.

The drawback here is the necessity for a global map to sustain the mapping between Maestro instances and consumer lists.